### PR TITLE
Update crypto command

### PR DESCRIPTION
**Pull request by matt-neal**
Remove specific hard-coded calls
Split crypto commands to their own file
Pull decorator logic to it's own file
Create a dict of crypto keys/names to search from
Update crypto to take in multiple params instead of none and be agnostic of separator type

### CODE REVIEW

File: commands.py [hunk 1]
```diff
@@ -1962,8 +1897,8 @@ def commit(message_dict):
 	send_message("The commit is running..")
 	log(f"Commited to Git for {user}")
 
-def _commit_thread(message):
-	result = os.system("commit.bat " + message)
+def _commit_thread():
+	result = os.system("commit.bat")
```
Comment (AcesFullOfKings) [inline commands.py]: Why did you remove the commit message?

File: crypto.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        if value == value:
+            return key
+
+    return "BTC"
+
+@is_command("Check all the crypto prices.")
+def crypto(message_dict):
+    user = message_dict["display-name"].lower()
+    message = [message_dict["message"]]
+    message_dict["suppress_log"] = True
```
Comment (AcesFullOfKings) [inline crypto.py]: I only added this "suppress_log" item to the input dict so that the individual crypto functions wouldn't unnecessarily log when called from the crypto function, but they would log when called directly. If suppress_log is set here and never modified then this function will always log, so this line can probably just be removed, along with the check below.

File: crypto_data.py [hunk 3]
```diff
@@ -0,0 +1,3760 @@
+crypto_data = {
```
Comment (AcesFullOfKings) [inline crypto_data.py]: Ah, if I'd known that you were asking for the list for RoboKaywee then I could've told you that the Coinbase API only accepts a few different coin types - only these pairs are supported https://api.pro.coinbase.com/products

File: crypto.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+from decorators      import is_command
+from crypto_data     import crypto_data
+
+def get_key(value):
+    for key, value in crypto_data.items():
+        if value == value:
+            return key
+
+    return "BTC"
```
Comment (AcesFullOfKings) [inline crypto.py]: If not found, this should either except or return None, which the calling function should handle.

File: commands.py [hunk 5]
```diff
@@ -1939,12 +1879,7 @@ def mycolour(message_dict):
 
 @is_command("Check all the crypto prices.")
 def crypto(message_dict):
-	user = message_dict["display-name"].lower()
-	message_dict["suppress_log"] = True
-	btc(message_dict)
-	eth(message_dict)
-	doge(message_dict)
-	log(f"Sent Crypto prices to {user}")
+	crypto_command.crypto(message_dict)
```
Comment (AcesFullOfKings) [inline commands.py]: I don't really see the reason to put this in a separate file - why don't you want it to stay in `commands.py`?

File: crypto.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+@is_command("Check all the crypto prices.")
+def crypto(message_dict):
+    user = message_dict["display-name"].lower()
+    message = [message_dict["message"]]
```
Comment (AcesFullOfKings) [inline crypto.py]: Did you mean to assign this into a list..? It causes the regex on line 22 to except as you can't do a regex on a list.

File: crypto.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    message_dict["suppress_log"] = True
+    
+    # Attempt to split the message and get each item into an array
+    try:
+        message = re.split(', |; |\s', message)
```
Comment (AcesFullOfKings) [inline crypto.py]: Do we need to use regex here? Regex is pretty slow so I usually use `message.split(" ")` in other commands. In addition, `message` contains the full chat message, so the first word is always "!crypto", which will fail to resolve in the dict. 

To get the command's arguments in other commands I use `message.split()[1:]`

File: crypto.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    # Fail gracefully if it's a single or empty message
+    except Exception as ex:
+        log(f"Exception in crypto.crypto: {str(ex)}")
+
+    # Check each item requested and make sure it's the short-code and not the full name
+    # Convert to short-code from name if needed
+    for item in message:
+        if item.length > 3:
```
Comment (AcesFullOfKings) [inline crypto.py]: * Why the length filter? 
* Some codes are shorter than 3 characters - I found 29 codes in my list is with len 2. 
* To get an item's length, use `len(item)` to match the rest of my code.

File: crypto.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            crypto = [get_key(item)]
```
Comment (AcesFullOfKings) [inline crypto.py]: * Are you expecting users to type the full name exactly? I'd suggest the arguments to the command should be the abbreviation, e.g. "!crypto btc" rather than "!crypto bitcoin".
* This assumes that all coin names are a single word (per line 22), but many coins have 2+ words in the name so would never be able to be looked up.
* Again, did you mean to make this a list? It causes the request on line 41 to fail if `item` is a list. Was this tested?

File: crypto.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            message[item] = crypto
```
Comment (AcesFullOfKings) [inline crypto.py]: I'd prefer to assign this to a second list rather than overwriting the original

File: crypto.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        if item.length > 3:
+            crypto = [get_key(item)]
+            message[item] = crypto        
+
+    # If the message is empty, return the default values
+    if message[0] == "":
+        message = ["BTC", "ETH", "DOGE"]
+
+    # Return the current value(s) of the requested crypto
+    for item in message:
```
Comment (AcesFullOfKings) [inline crypto.py]: We're iterating through the message twice which is inefficient - can we merge these loops into one?

File: crypto.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Check each item requested and make sure it's the short-code and not the full name
+    # Convert to short-code from name if needed
+    for item in message:
+        if item.length > 3:
+            crypto = [get_key(item)]
+            message[item] = crypto        
+
+    # If the message is empty, return the default values
+    if message[0] == "":
```
Comment (AcesFullOfKings) [inline crypto.py]: By definition, `message[0]` is always "!crypto", so this will never pass. Even if it weren't and the list was somehow empty, [] != "" so this would still never pass

File: crypto.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+from crypto_data     import crypto_data
+
+def get_key(value):
+    for key, value in crypto_data.items():
+        if value == value:
+            return key
+
+    return "BTC"
+
+@is_command("Check all the crypto prices.")
```
Comment (AcesFullOfKings) [inline crypto.py]: This shouldn't be decorated with @is_command as it's not callable directly from chat. The decorator on commands.crypto is correct, but as this is a sub-function is shouldn't be decorated and should really have a leading underscore on the function name.

File: crypto.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
+import requests
+import re
+
+from decorators      import is_command
+from crypto_data     import crypto_data
+
+def get_key(value):
+    for key, value in crypto_data.items():
+        if value == value:
```
Comment (AcesFullOfKings) [inline crypto.py]: This is always True..? Was this tested?

File: crypto.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
+import requests
+import re
+
+from decorators      import is_command
+from crypto_data     import crypto_data
+
+def get_key(value):
+    for key, value in crypto_data.items():
```
Comment (AcesFullOfKings) [inline crypto.py]: If you really need a two-way lookup then I'd suggest probably a list of tuples is better than a dict. Although as below I'm not sure why we need to look up a code based on a full coin name.

File: crypto.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    user = message_dict["display-name"].lower()
+    message = [message_dict["message"]]
+    message_dict["suppress_log"] = True
+    
+    # Attempt to split the message and get each item into an array
+    try:
+        message = re.split(', |; |\s', message)
+    # Fail gracefully if it's a single or empty message
+    except Exception as ex:
+        log(f"Exception in crypto.crypto: {str(ex)}")
```
Comment (AcesFullOfKings) [inline crypto.py]: This should also `return False`, and probably also put a message in chat saying there was an error - see other commands for my error handling.

File: decorators.py [hunk 17]
```diff
@@ -0,0 +1,14 @@
+def is_command(description=""):
```
Comment (AcesFullOfKings) [inline decorators.py]: Again, I'm not really sure why this needs to be in its own file when there's only one short function.

File: crypto.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
+import requests
+import re
+
+from decorators      import is_command
```
Comment (AcesFullOfKings) [inline crypto.py]: This can be removed

**Review by AcesFullOfKings — COMMENTED** [review]
I'm not sure that having the command as "!crypto bitcoin ethereum [..] litecoin" is better than just having a dedicated command for each coin. If there's demand for commands for further coins then I'll gladly add e.g. !litecoin etc. But in practise I can't see someone in chat typing out a list of the full names of all the coins they want to see. In particular, even if chatters were going to type out a full list, I'd think they would prefer to use the short names as parameters, i.e. "!crypto eth btc ltc". 

I think the idea of being able to retrieve arbitrary coins is great, but this implementation needs a bit of thought and especially needs testing before I can merge it. 

If I were designing this, I'd have "!crypto" which would behave as currently, showing the values for the three most relevant coins to Kaywee's chat; or alternatively it could accept a single argument, for example "!crypto ltc", to look up a specific coin through the API, and put a fail message in chat if it doesn't work. That method avoids the need for a local dict of all coins and also provides a simpler interface; I think most commands should prefer being simple to use over being functionally broad, as in practise chatters want to write as little as possible to get the bot to work.

File: commands.py (hunk 1 above)
Comment (matt-neal) [reply commands.py]: I'm not sure where or how that disappeared lol, will fix.

File: commands.py (hunk 5 above)
Comment (matt-neal) [reply commands.py]: Because I'm going to start parsing the commands out of `commands.py` as well. It's exceptionally unwieldy to maintain already and is still being added to. We're making it perfectional. ;)

**Comment by matt-neal** [comment]
> I'm not sure that having the command as "!crypto bitcoin ethereum [..] litecoin"...

This implementation doesn't just accept the whole-name set of arguments. This method now accepts:
`No args: returns btc, eth, and doge` (previous behavior)
`Single arg in string or short-code: returns that value` (updated to take in a single param)
`Multiple args without caring too much about separator: returns those values`

So the functionality is the same as before but also improved to not be so persnickety.

File: crypto_data.py (hunk 3 above)
Comment (matt-neal) [reply crypto_data.py]: I can shorten the dict to the 74 codes allowed then, easy.

File: crypto.py (hunk 4 above)
Comment (matt-neal) [reply crypto.py]: Yeah, I did a lazy pass here to return BTC to not handle the exception. Definitely can fix.

File: crypto.py (hunk 6 above)
Comment (matt-neal) [reply crypto.py]: Ah, yes, I forgot to remove the array notation since I'm turning the string into an array with split.

File: crypto_data.py (hunk 3 above)
Comment (AcesFullOfKings) [reply crypto_data.py]: I still think we don't need the dict. We can just try the coinbase request and see if it works - if not then fail gracefully with a chat mesage. That way it's futureproof - we don't have to keep a dict up to date - and also simpler.

File: crypto.py (hunk 7 above)
Comment (matt-neal) [reply crypto.py]: Large regex functionality is slow, but minor parsing regex isn't any slower than `string.split()`, at least perceptibly.
Really, I plan on pulling out all of the `split()` calls into something like this, just to keep from having so many failures if someone adds a comma or semi-colon instead of just using spaces

File: crypto.py (hunk 8 above)
Comment (matt-neal) [reply crypto.py]: The length filter is to check if it's a name or short-code, but I probably should improve that logic by doing something more along the lines of `if crypto_data[item]` which would be better logic and also let me quickly handle failure cases rather than running the reverse lookup too.

File: crypto.py (hunk 7 above)
Comment (AcesFullOfKings) [reply crypto.py]: I prefer string.split() as it's simpler to read. 

In practise, I doubt that anyone will use commas or semicolons as that's not how chatbots work - and even if they do try something like that, it's just a syntax error and it's fine for the command to return a failure message to chat, or even for the bot to just ignore them. It doesn't need converting to regex project-wide.

File: crypto.py (hunk 8 above)
Comment (AcesFullOfKings) [reply crypto.py]: The length would only be a valid check if all of the codes were shorter than all of the names, which is not the case: the longest code I found is 7 characters, and the shortest name is 2. 

I think allowing either a code or a name is overcomplicating it - it should allow only a code. That way it's far simpler. And in practise I don't think people are going to type out the full name, especially if they have the option of using the code.

**Comment by AcesFullOfKings** [comment]
I really think keeping it simple should be a priority. It doesn't need to try to handle every permutation of names and codes and accepting multiple arguments. I think a simple "!crypto [\<code>]" would be far preferable, which accepts 0 or 1 coin code. If zero are provided it behaves as current, while if 1 code is provided it tries to look up that coin's price via coinbase, and if it fails it puts a failure mesasge in chat. That way we don't need to maintain our own dict, it can be kept as a simple command function in commands.py, and we don't have unnecessary complexity.

File: commands.py [hunk 19]
```diff
@@ -32,7 +34,7 @@ def inner(func, description=description):
 
 """
 Each @is_command function is a command (!!), callable by sending "!<function_name>" in chat.
-All replies will be sent in the bot's colour, using /me unless specified otherwise.
+All replies will be sent in the bots colour, using /me unless specified otherwise.
```
Comment (AcesFullOfKings) [inline commands.py]: This has snuck back in!

File: commands.py [hunk 20]
```diff
@@ -1940,11 +1895,24 @@ def mycolour(message_dict):
...
 	user = message_dict["display-name"].lower()
-	message_dict["suppress_log"] = True
-	btc(message_dict)
-	eth(message_dict)
-	doge(message_dict)
-	log(f"Sent Crypto prices to {user}")
+	message = message_dict["message"]
+
+	# If the message is empty, return the default values
+	if message[1] == "":
```
Comment (AcesFullOfKings) [inline commands.py]: `message[1]` here refers to the string, so it's going to compare against the second character. I think you're expecting it to be a list.

In other commands I check for arguments by wrapping the `message.split(" ")[1]` in a try/except, and if it excepts IndexError (i.e. no item in position 1) I set the default arguments. Perhaps you could move `crypto_codes` up do it like that, but up to you. 

At the moment you're also overwriting `crypto_codes` below even if it does get set here.

File: commands.py [hunk 21]
```diff
@@ -1940,11 +1895,24 @@ def mycolour(message_dict):
...
+	for item in crypto_codes:
+		try:
+			result = requests.get(f"https://api.coinbase.com/v2/prices/{item.upper()}-USD/spot").json()
+			value = float(result["data"]["amount"])
+		except Exception as ex:
+			log(f"Exception in crypto.crypto: {str(ex)}")
+			send_message(f"{item.upper()} is not currently available via coinbase")
+			return False
+
+		send_message(f"{item.upper()} is currently worth ${value:,}")
```
Comment (AcesFullOfKings) [inline commands.py]: Whenever I send a message I also log, so could you add a call to `log("some log string")` here please, preferably recording both the user and the response.

File: commands.py [hunk 22]
```diff
@@ -13,6 +13,8 @@
 from multiprocessing import Process
 from james           import seconds_to_duration, timeuntil
 from contextlib      import suppress
+import crypto as crypto_command
+from decorators		 import is_command
```
Comment (AcesFullOfKings) [inline commands.py]: As we've removed the other files, we can also remove the imports (which would fail anyway).

File: commands.py [hunk 23]
```diff
@@ -1940,11 +1895,24 @@ def mycolour(message_dict):
...
+	# If the message is empty, return the default values
+	if message[1] == "":
+		crypto_codes = ["BTC", "ETH", "DOGE"]
+
+	# Return the current value(s) of the requested crypto
+	crypto_codes = message.split(" ")[1:]
+	for item in crypto_codes:
+		try:
+			result = requests.get(f"https://api.coinbase.com/v2/prices/{item.upper()}-USD/spot").json()
+			value = float(result["data"]["amount"])
```
Comment (AcesFullOfKings) [inline commands.py]: Could we also round this to 2dp please, as it somes with many decimal places which we don't really want in chat.

**Review by AcesFullOfKings — CHANGES_REQUESTED** [review]
Thanks for this! Looks great. I've added a couple of comments but it's nearly ready to merge. 

Could you also test this by creating a `message_dict` with the expected values, and overriding the `send_mesasge` and `log` functions in a test file (presumably just print the message to console) and then calling the function with various different `message_dict["message"]` values (which holds the message as received from chat, e.g. "!crypto btc" etc). Then you can check there are no errors and the output is as expected.

File: commands.py [hunk 24]
```diff
@@ -1371,53 +1373,6 @@ def wordoftheday(message_dict):
 	send_message(f"{tag}The Spanish Word of the Day is \"{spa}\", which means \"{eng}\"")
 	log(f"Sent word of the Day to {user}: {spa} means {eng}")
 
-@is_command("Show the price of bitcoin")
-def btc(message_dict):
```
Comment (AcesFullOfKings) [inline commands.py:1385]: Oh, I think we should also keep the btc, eth, and doge commands as Kaywee uses them often. And I think having a special command for the 2-3 coins of most interest to the chat is a good idea, bc it reduces the length of the command from e.g. `!crypto eth` to just `!eth`. Then for any other less-common coins chatters can still use `!crypto <coin>`, or for the three common ones they can still use the `!crypto` shortcut.

If you want to edit them so they use fewer lines of code then that's fine, but I'd like to keep the commands.

File: commands.py (hunk 19 above)
Comment (matt-neal) [reply commands.py]: Yeah, I think there were some artifacts from me changing origins on my branch.

File: commands.py (hunk 23 above)
Comment (matt-neal) [reply commands.py]: Thoughts on doing 4 instead of 2 in events someone looks up something that is under a penny value?

File: commands.py (hunk 20 above)
Comment (matt-neal) [reply commands.py]: Yep, I was also stuck on a work bug so my brain scrambled this logic.

File: commands.py (hunk 23 above)
Comment (AcesFullOfKings) [reply commands.py]: I don't know if it's possible to do significant figures.. is that possible? 

If not it could be worth adding a `dp = 2 if value >= 0.01 else 4`. 

Is this likely to come up? I think anyone in chat checking a coin other than eth/btc/doge is going to be quite rare, and I think by definition there's little interest in coins with sub-penny values. It would be fine to ignore that edge case and accept that the value would come back as 0.00.

**Comment by matt-neal** [comment]
> Thanks for this! Looks great. I've added a couple of comments but it's nearly ready to merge.
> 
> Could you also test this by creating a `message_dict` with the expected values, and overriding the `send_mesasge` and `log` functions in a test file (presumably just print the message to console) and then calling the function with various different `message_dict["message"]` values (which holds the message as received from chat, e.g. "!crypto btc" etc). Then you can check there are no errors and the output is as expected.

I'm on my Mac that won't let me `pip install requests` for some who knows what reason. Here is the test snippet I wrote that would likely be easier for you to run locally and check for me.

```Python
import requests
import os

def send_message(msg):
	print ("message: " + msg)
  
def log(msg):
	print ("log: " + msg)

def eth(message_dict):
	new_message_dict = message_dict
	new_message_dict["message"] = "!eth eth"
	crypto(new_message_dict)

def btc(message_dict):
	new_message_dict = message_dict
	new_message_dict["message"] = "!btc btc"
	crypto(new_message_dict)

def eth(message_dict):
	new_message_dict = message_dict
	new_message_dict["message"] = "!doge doge"
	crypto(new_message_dict)

def crypto(message_dict):
	user = message_dict["display-name"].lower()
	message = message_dict["message"]

	try:
		crypto_codes = message.split(" ")[1:]
	except:
		crypto_codes = ["BTC", "ETH", "DOGE"]

	for item in crypto_codes:
		try:
			if (item.upper() == "DOGE"):
				result = requests.get(f"https://sochain.com/api/v2/get_price/{item.upper()}/USD").json()
			else:
				result = requests.get(f"https://api.coinbase.com/v2/prices/{item.upper()}-USD/spot").json()
			value = float(result["data"]["amount"])
		except Exception as ex:
			log(f"Exception in crypto: {str(ex)}")
			send_message(f"{item.upper()} is not currently available via coinbase")
			return False

		send_message(f"{item.upper()} is currently worth ${round(value, 4):,}")
		log(f"Sent {item.upper()} of ${round(value, 4)} to {user}")
    
base_test_dict = {"user": "MoldarOW", "message": "!crypto"}
btc_test_dict = {"user": "MoldarOW", "message": "!btc"}
eth_test_dict = {"user": "MoldarOW", "message": "!eth"}
doge_test_dict = {"user": "MoldarOW", "message": "!doge"}
valid_test_dict = {"user": "MoldarOW", "message": "!crypto BTC ETH doge BnB aDa"}
invalid_test_dict = {"user": "MoldarOW", "message": "!crypto asdpofiuwqe"}
invalid_test_dict_2 {"user": "MoldarOW", "message": "!crypto 158901u234 1283512p03895 750y1hufgasd70ya"}

test_dict_array = [base_test_dict, valid_test_dict, invalid_test_dict, invalid_test_dict_2]

for item in test_dict_array:
	crypto(item)
  
btc(btc_test_dict)
eth(eth_test_dict)
doge(doge_test_dict)
btc(eth_test_dict)
eth(doge_test_dict)
doge(btc_test_dict)
 ```

File: commands.py [hunk 25]
```diff
@@ -1937,14 +1890,48 @@ def mycolour(message_dict):
...
+	crypto(new_message_dict)
+
+@is_command("Show the price of bitcoin")
+def btc(message_dict):
+	new_message_dict = message_dict
+	new_message_dict["message"] = "!btc btc"
+	crypto(new_message_dict)
+
+@is_command("Show the price of dogecoin")
+def eth(message_dict):
```
Comment (AcesFullOfKings) [inline commands.py]: Looks like this should be `def doge`?

And the messages should begin with !crypto instead of the coin name.

Otherwise the rest of this looks good. I'll test it tomorrow and merge in once the command name is fixed. 

Thanks for the effort! :D

**Comment by AcesFullOfKings** [comment]
Is this waiting for me to do something? Just checking we're not both waiting for each other :)

**Comment by AcesFullOfKings** [comment]
The tests wouldn't run as-is but I modified the code as below and got the tests working. If you commit this I can merge it in :)

<code>

	def eth(message_dict):
		new_message_dict = message_dict
		new_message_dict["message"] = "!crypto eth"
		crypto(new_message_dict)

	def btc(message_dict):
		new_message_dict = message_dict
		new_message_dict["message"] = "!crypto btc"
		crypto(new_message_dict)

	def doge(message_dict):
		new_message_dict = message_dict
		new_message_dict["message"] = "!crypto doge"
		crypto(new_message_dict)

	def crypto(message_dict):
		user = message_dict["display-name"].lower()
		message = message_dict["message"]

		try:
			crypto_codes = message.split(" ")[1:]
		except:
			crypto_codes = ["BTC", "ETH", "DOGE"]

		for item in crypto_codes:
			item = item.upper()
			try:
				if item == "DOGE":
					result = requests.get(f"https://sochain.com/api/v2/get_price/{item}/USD").json()
					value = float(result["data"]["prices"][0]["price"])
				else:
					result = requests.get(f"https://api.coinbase.com/v2/prices/{item}-USD/spot").json()
					value = float(result["data"]["amount"])
			except Exception as ex:
				log(f"Exception in crypto: {str(ex)}")
				send_message(f"{item} is not currently available via coinbase")
				return False

			send_message(f"{item} is currently worth ${round(value, 4):,}")
			log(f"Sent {item} of ${round(value, 4)} to {user}")
</code>

**Review by AcesFullOfKings — APPROVED** [review]
